### PR TITLE
FIX: monthly upload, running successfully locally

### PR DESCRIPTION
**Pull request by violetbrina**

### CODE REVIEW

File: cpg_infra/billing_aggregator/aggregate/utils.py [hunk 1]
```diff
@@ -797,15 +800,50 @@ def get_unit_for_batch_resource_type(batch_resource_type: str) -> str:
...
+    elif content_type in ('application/octet-stream', 'text/plain'):
+        print('Text data found')
+        request_data = json.loads(request.data)
+    elif content_type == 'application/x-www-form-urlencoded':
+        print('Encoded Form')
+        request_data = request.form
+    else:
+        raise ValueError(f'Unknown content type: {content_type}')
+
+    if 'attributes' in request_data and (
```
Comment (illusional) [inline cpg_infra/billing_aggregator/aggregate/utils.py:827]: Nitpick, but you don't need this `"attributes" in requests_data` as the below is defensive (`request_data.get('attributes')` won't crash)

File: cpg_infra/billing_aggregator/aggregate/utils.py [hunk 2]
```diff
@@ -797,15 +800,50 @@ def get_unit_for_batch_resource_type(batch_resource_type: str) -> str:
...
-    if request:
-        print(request)
-        # return request.params['start'], request.params['end']
-    return None, None
+    if not request:
+        return None, None
+
+    content_type = request.content_type
+    if request.method == 'GET':
+        print('GET request, using args')
```
Comment (illusional) [inline cpg_infra/billing_aggregator/aggregate/utils.py:813]: Can you switch the prints to logging, also could be worth logging the args in these blocks below too

File: cpg_infra/billing_aggregator/monthly_aggregate/main.py [hunk 3]
```diff
@@ -32,10 +34,63 @@ def get_bigquery_client() -> bq.Client:
...
 
-def from_request(*args, **kwargs):
+def get_invoice_month_from_request(
+    request: Request,
+) -> str | None:
+    """
+    Get the invoice month from the cloud function request.
+    """
+    if not request:
+        print('No request found')
```
Comment (illusional) [inline cpg_infra/billing_aggregator/monthly_aggregate/main.py:44]: Ditto with print -> logging + logging args

File: cpg_infra/billing_aggregator/monthly_aggregate/main.py [hunk 4]
```diff
@@ -32,10 +34,63 @@ def get_bigquery_client() -> bq.Client:
...
+@functions_framework.http
+def from_request(request: Request):
     """Entrypoint for cloud functions, run always as default (previous month)"""
-    asyncio.new_event_loop().run_until_complete(
-        process_and_upload_monthly_billing_report(None)
+
+    try:
+        invoice_month = get_invoice_month_from_request(request)
+    except ValueError:
+        invoice_month = None
```
Comment (illusional) [inline cpg_infra/billing_aggregator/monthly_aggregate/main.py:90]: Would be worth logging the ValueError here and indicating that you're setting it to None (the default)

File: cpg_infra/billing_aggregator/monthly_aggregate/main.py [hunk 5]
```diff
@@ -32,10 +34,63 @@ def get_bigquery_client() -> bq.Client:
...
+    elif 'message' in request_data and 'data' in request_data.get('message'):
+        try:
+            request_data = json.loads(b64decode(request_data['message']['data']))
+        except Exception as exp:
+            raise exp
+
+    print(request_data)
+
+    if request_data and 'invoice_month' in request_data:
+        invoice_month = request_data.get('invoice_month')
```
Comment (illusional) [inline cpg_infra/billing_aggregator/monthly_aggregate/main.py:76]: Nitpick: you can early return here

File: cpg_infra/billing_aggregator/monthly_aggregate/main.py [hunk 6]
```diff
@@ -68,25 +123,31 @@ async def process_and_upload_monthly_billing_report(invoice_month: str = None):
...
     Creates the batch_update the user has access to.
     Load pre-authorized user credentials from the environment.
     TODO(developer) - See https://developers.google.com/identity
     for guides on implementing OAuth2 for the application.
     """
 
-    creds, _ = google.auth.default(
-        scopes=['https://www.googleapis.com/auth/spreadsheets']
-    )
+    year = invoice_month[:4]
```
Comment (illusional) [inline cpg_infra/billing_aggregator/monthly_aggregate/main.py:144]: The hot trend is to be more assertive with Python where you expect the shape of specific data

```suggestion
    assert len(invoice_month) == 6
    year = invoice_month[:4]
```

File: cpg_infra/billing_aggregator/monthly_aggregate/main.py [hunk 7]
```diff
@@ -145,4 +207,4 @@ def get_billing_data(invoice_month: str) -> DataFrame:
     logging.getLogger('urllib3').setLevel(logging.WARNING)
     event_loop = asyncio.new_event_loop()
 
-    event_loop.run_until_complete(process_and_upload_monthly_billing_report())
+    # event_loop.run_until_complete(process_and_upload_monthly_billing_report(None))
```
Comment (illusional) [inline cpg_infra/billing_aggregator/monthly_aggregate/main.py:210]: Is this intentionally commented out? Should still work in cloud function, but you couldn't run it locally right?

**Review by illusional — APPROVED** [review]

File: cpg_infra/billing_aggregator/monthly_aggregate/main.py (hunk 7 above)
Comment (violetbrina) [reply cpg_infra/billing_aggregator/monthly_aggregate/main.py]: Yup unintentional

**Pull request merged**